crates/core/src/operations/merge/mod.rs

```diff
@@ -1427,11 +1439,62 @@ async fn execute(
     });
 
     let operation_count = DataFrame::new(state.clone(), operation_count);
+
+    if should_cdc {
+        // Create a dataframe containing the CDC deletes which are present at this point
+        change_data.push(
+            operation_count
+                .clone()
+                .filter(col(DELETE_COLUMN))?
+                .select(write_projection.clone())?
+                .with_column(crate::operations::cdc::CDC_COLUMN_NAME, lit("delete"))?,
+        );
+    }
+
     let filtered = operation_count.filter(col(DELETE_COLUMN).is_false())?;
 
-    let project = filtered.select(write_projection)?;
-    let merge_final = &project.into_unoptimized_plan();
+    if should_cdc {
+        debug!("The merge should triggere a CDC tracking, computing pre/insert/postimage datasets");
+        let cdc_projection = filtered.clone().filter(col(OPERATION_COLUMN).not_eq(
+            // This is a copy operation, but I'm not sure how to turn that enum into an int
+            lit(5),
+        ))?;
+
+        change_data.push(
+            cdc_projection
+                .clone()
+                .filter(
+                    col(SOURCE_COLUMN)
+                        .is_true()
+                        .and(col(TARGET_COLUMN).is_null()),
+                )?
+                .select(write_projection.clone())?
+                .with_column(CDC_COLUMN_NAME, lit("insert"))?,
```

Comment:
Why do we create inserts? Isn't it cheaper to let the cdf_reader calculate what was inserted?

Reply:
@rtyler I think we can add one optimization here, if we see it's insert only. So there is only a when_not_matched_insert_all statement, then we could avoid writing CDC as a whole.